File: 1103. Distribute Candies to People/Solution.py
import math
from typing import List


class Solution:
    def distributeCandies_MK1(self, candies: int, num_people: int) -> List[int]:
        '''
        BF.
        '''
        i = 1
        ret = [0] * num_people
        while candies > i:
            ret[(i - 1) % num_people] += i
            candies -= i
            i += 1
        ret[(i - 1) % num_people] += candies
        return ret

    def distributeCandies_MK2(self, candies: int, num_people: int) -> List[int]:
        '''
        Math.
        '''
        n = int(math.sqrt(2 * candies + 0.25) - 0.5)
        rows, cols = divmod(n, num_people)
        ret = [0] * num_people
        for i in range(num_people):
            ret[i] += (i + 1) * rows + rows * (rows - 1) * num_people // 2
            if i < cols:
                ret[i] += 1 + i + rows * num_people
        ret[cols] += candies - n * (n + 1) // 2
        return ret

File: 1103. Distribute Candies to People/Solution.java
class Solution {
    public int[] distributeCandies(int candies, int num_people) {
        int n = (int)(Math.sqrt(2 * candies + 0.25) - 0.5);
        int rows = n / num_people;
        int cols = n % num_people;
        int[] ret = new int[num_people];
        for (int i = 0; i < num_people; i++) {
            ret[i] = (int)((rows - 1) * rows * 0.5 * num_people) + rows * (i + 1);
            if (i < cols)
                ret[i] += rows * num_people + i + 1;
        }
        ret[cols] += candies - (1 + n) * n * 0.5;
        return ret;
    }
}
